Cabal/doc/developing-packages.rst

```diff
@@ -2184,14 +2184,27 @@ system-dependent values for these fields.
        :pkg-field:`other-modules`, :pkg-field:`library:exposed-modules` or
        :pkg-field:`executable:main-is` fields.
 
+.. pkg-field:: hs-source-dir: directory list
+    :deprecated: 2.0
+    :removed: 3.0
+    :default: ``.``
+
+    Root directories for the module hierarchy.
+
+    Deprecated in favor of :pkg-field:`hs-source-dirs`.
+
 .. pkg-field:: hs-source-dirs: directory list
 
     :default: ``.``
 
     Root directories for the module hierarchy.
 
-    For backwards compatibility, the old variant ``hs-source-dir`` is
-    also recognized.
+    .. note::
+
+      Components can share source directories but modules will be compiled if
+      they can be found there, i.e., if a library and an executable share a
+      source directory and the executable depends on the library's ``Foo``
+      module it will be recompiled even if the library has already been built.
```

Comment:
You could add a note that internal libraries are a way to solve that problem.

Reply:
Isn't the real solution not to share a tree with dependencies? Or would internal libraries's modules in the src dir not be recompiled?

Reply:
I'm not sure what you mean by "sharing a tree with dependencies"?

What you would do is add the internal library as a build-depends to the components where you want to share modules so then they only get compiled once, namely when the internal lib is built but you can still use them all over.

Reply:
The tree I'm talking about is a file hierarchy. No matter what you do if a module you import is in your source dir you'll compile it again. Even if a dependency, internal lib or no, already built it.

Reply:
@DanielG, I'm not clear on the problem you suppose would be solved here using internal libraries? I can imagine people might share a source dir if they wanted multiple parts depending on something without exposing it but the reason for sharing the dirs isn't really relevant to the note. The note's just supposed to point out a gotcha people might run into. Another way you could end up in this situation is simply desiring a simpler repository setup, with a single src directory for all your code.

Reply:
I think I missed the fact that you were talking specifically about (public) libraries already, so nevermind.

I was talking about that other common problem where you want your test suite to have access to internals without exporting them in the public library but that doesn't really fit here, you're right.